### ISSUE TITLE
Txrep: Canonical human-readable representation of Stellar transactions

### ISSUE THREAD

**Issue by stanford-scs** [stellar/stellar-protocol#159]
## Preamble

    SEP: Not assigned
    Title: Txrep: human-readable low-level representation of Stellar transactions
    Author: David Mazières
    Status: Draft
    Created: 2018-08-31

## Simple Summary

Txrep is a human-readable representation of Stellar transactions that functions like an assembly language for XDR.

## Abstract

This document specifies *txrep*, a human-readable format for Stellar transactions. Txrep is unambigous and machine-parsable. Binary XDR transactions can be decompiled into txrep format and recompiled to the exact same binary bytes. Txrep is designed to be parsed directly into data structures generated by XDR compilers, ideally by the very XDR code generated by those compilers.

## Motivation

Bug reports, test vectors, semantic specifications in CAP documents, and Stellar documentation all need a way to talk concretely about the contents of transactions. Without a canonical textual format, different documents will each devise their own way of describing transactions, sometimes ambiguously, making it hard to relate information from multiple sources.

Furthermore, advanced Stellar users and devlopers need a way to craft arbitrary transactions, including potentially invalid ones for test cases. This functionality is currently available from the [Stellar laboratory transaction builder](https://www.stellar.org/laboratory/#txbuilder), but that is not a good solution for high security, as it requires a browser (and in practice requires trusting an HTTPS certificate). Moreover, the transactions one builds in a web browser cannot easily be audited before signing, or described in documentation, or scripted, or placed under version control.

Finally, "dumb" ink-on-paper contracts may need to specify Stellar transactions unambiguously. For example, a legal contract promising to deliver tokens with a lock-up period in exchange for a wire transfer may want to specify the exact lock-up mechanism used through a human readable description of the Stellar transactions involved. Using txrep, this description can be unambiguous.

## Specification

A txrep file consists of a number of lines, each describing the value of a field. We describe the line format using a simple BNF-like notation in which brackets (\[...\]) indicate optional contents, pipe (|) indicates alternatives, and an asterisk (\*) indicates zero or more repetitions of the previous symbol. Literal text (e.g., `:`, `.`, `[`, `]`, `._present`, `.len`, or `_`) indicates the occurence of those specific characters in the txrep source.

Each line of txrep has the following format:

> *line* = *field* `:` SP\* *value* \[*comment*\] LF | `:` *comment* LF | LF
> 
> *comment* = zero or more characters other than LF
> 
> SP = space (ASCII 32)
> 
> LF = newline (ASCII 10)

Blank lines are allowed, and any line starting with a colon (`:`) is a full-line comment. The *field* and *value* formats are described below.

### Fields

A field has the following syntax:

> *field* = *tag* *selector*\* \[*pseudoselector*\]
> 
> *selector* = `.` *tag* | `[` *integer* `]`
> 
> *tag* = *letter* *wordchar*\*
> 
> *pseudoselector* = `.` `_inner`\* `_present` | `.len`
> 
> *integer* = A decimal integer
> 
> *letter* = Any letter
> 
> *wordchar* = Any letter, any digit, or `_`

Disregarding *pseudoselector*, each *field* names a field in an [XDR `TransactionEnvelope` data structure](https://github.com/stellar/stellar-core/blob/master/src/xdr/Stellar-transaction.x#L351). The name is generated by joining XDR field names with a period (`.`), similar to the syntax used for accessing nested fields in C++ or go representations of XDR data structures. Array elements (for both fixed- and variable-length arrays) are indexed using square brackets. As in C, arrays are 0-based. Pseudoselectors of the form `_inner_present` (or `_inner_inner_present`, etc.) are reserved for fields of type pointer to pointer. Such types do not appear in the current version of `TransactionEnvelope`, but might in the future, or txrep might be applied to other XDR data types that contain such nested pointers.

As an example, the *field* `tx.timeBounds.minTime` names the `minTime` field in the `tx` field of the `TransactionEnvelope` structure, which looks like this:

``` c
struct TimeBounds {
    uint64 minTime;
    uint64 maxTime; // 0 here means no maxTime
};
struct Transaction {
    /* ... */
    TimeBounds* timeBounds;
    Operation operations<100>;
    /* ... */
};
struct TransactionEnvelope {
    Transaction tx;
    DecoratedSignature signatures<20>;
};
```

Pointers and variable-length arrays use pseudoselectors to describe their state.

Pointers use the `._present` pseudoselector with value `true` or `false` to indicate that the field is present or NULL, respectively. For example, a transaction with timebounds might be specified like this (using comments to annotate the times):

    tx.timeBounds._present: true
    tx.timeBounds.minTime: 1535756672 (Fri Aug 31 16:04:32 PDT 2018)
    tx.timeBounds.maxTime: 1567292672 (Sat Aug 31 16:04:32 PDT 2019)

A transaction without timebounds would contain this line:

    tx.timeBounds._present: false

For fields of type pointer-to-pointer, the pseudoselector `._present` describes whether the outermost pointer is NULL or not. If `._present` is `true`, then `._inner_present` describes whether the nested pointer is NULL or not. `_inner` can be repeated as many times as needed for arbitrarily nested pointers.

Variable-length arrays use the pseudo-selector `.len` with an integer value to indicate the number of elements of the variable-length array. Only indices from 0 to len-1 are meaningful. For example, a transaction with one operation might look like this:

    tx.operations.len: 1
    tx.operations[0].sourceAccount._present: false
    tx.operations[0].body.type: PAYMENT
    ...

### Values

Most XDR types are rendered using C syntax. Specifically:

  - All integers (signed and unsigned, 32- and 64-bit) are represented as C integers (decimal by default, but prefix `0x` can be used for hex and `0` for octal).

  - `bool` values are `true` or `false`

  - Enums are represented by the bare keyword of the value. They cannot be specified numerically (but omitting a line for a field is the same as specifying 0).

  - `string` values are represented as double-quoted interpreted string literals, in which non-ASCII bytes are represented with hex escapes (`"\xff"`), the `"` and `\` characters can be escaped with another `\` (e.g., `"\\"`), and `\n` designates a newline.

  - `opaque` values are represented as an unquoted hexadecimal string (using lower-case case `a`...`f`)

A few aggregate values are special-cased:

  - The `AssetCode` field in `Assets` is rendered as ASCII bytes (not surrounded by double-quotes). Non-printable characters and non-ASCII bytes (including space, to avoid a space being interpreted as part of a comment) are hex escaped (`\xff`), while backslash is self-escaped (`\\`).

  - `PublicKey` and `SignerKey` are rendered as unquoted strings in strkey format, described below.

Any fields in the XDR `TransactionEnvelope` structure that are not specified in a txrep description are to be interpreted as zero (for numeric values, enums, and fixed-length `opaque`) and zero-length (for strings, variable-length arrays, and variable-length `opaque`).

### Strkey format

Strkey provides a compact ASCII format for ED25519 public keys, ED25519 private keys (also known as seeds), pre-authorized transaction hashes, and hash-x signers (which provide signing authority upon revelation of a SHA-256 preimage). Each of these four types has a corresponding version byte, which determines the first character of the strkey encoding:

| Key type                | Version byte | First char |
| ----------------------- | ------------ | ---------- |
| STRKEY\_PUBKEY\_ED25519 | 6            | G          |
| STRKEY\_SEED\_ED25519   | 18           | S          |
| STRKEY\_PRE\_AUTH\_TX   | 19           | T          |
| STRKEY\_HASH\_X         | 23           | X          |

The following steps transform a binary key into a strkey:

1.  Prepend the appropriate version byte to the binary key (so, e.g., a 32-byte ED25519 public key becomes 33 bytes).

2.  Compute a 16-bit CRC16 checksum of the combined version byte and binary key (using polynomial x<sup>16</sup> + x<sup>12</sup> + x<sup>5</sup> + 1). Append the two-byte checksum to the version byte and binary key (e.g., producing a 35-byte quantity for an ED25519 public key).

3.  Encode the concatenated Version-byte, Binary-key, and CRC16 using [RFC4648 base-32 encoding](https://tools.ietf.org/html/rfc4648#section-6).

The strkey format is already widely used for Stellar. Example imlementations include these in [go](https://github.com/stellar/go/blob/master/strkey/main.go), [C++](https://github.com/stellar/stellar-core/blob/dd7f38dba232e20ae75a95641857c0b7ed0252e1/src/crypto/StrKey.cpp), and [JavaScript](https://github.com/stellar/js-stellar-base/blob/master/src/strkey.js).

### Normalized txrep

Fields in txrep are specified in an order-independent way. If a field appears twice, the second value overrides the first. This allows one to update a txrep-format transaction by appending lines to a file. However, in some cases, such as when disassembling binary transactions, it is useful to transform transactions into normalized form, for instance so two transactions can be more easily compared, or so users inspecting a transaction see a more predictable format.

*Normalized txrep* format is a *txrep* format with the following additional restrictions:

  - Every field and pseudofield in the the binary XDR transaction must appear exactly once in the description. Extraneous fields or fields that do not appear because of NULL pointers or incompatible union discriminants must not appear.

  - Fields in structs and unions must appear in the exact order they appear in the XDR file, which is also the order in which they are marshaled for XDR binary format.

  - Pseudo-fields must appear immediately before the values they affect. In particular, the `ptr._present: true` field must appear immediately before the value of `ptr`, and `vector.len` must appear immediately before `vector[0]`.

## Rationale

Txrep is designed to be easy to read into and generate from structures created by XDR compilers using code generated by XDR compilers. Writing txrep simply requires printing each field and value in an XDR structure instead of rendering it in binary format. Txrep can also trivially be read into a key-value map, as almost every language has functions to read files line-by-line and break strings at a specific character (to separate field and value at `:`). The map can then be consulted in XDR deserialization routines to read txrep.

Tailoring txrep specifically to XDR means it does not introduce a second native representation for transactions. Third party libraries for other popular formats such as JSON and YAML generally cannot be parsed directly into a `TransactionEnvelope` structure generated by an XDR compiler, in part because of XDR's tagged unions. Moreover, the ability to parse and generate txrep with nothing more than an XDR implementation reduces the attack surface of programs that process txrep.

Because each line of txrep is entirely self-contained, one can excerpt any subset of a transaction with no ambiguity. Because txrep takes the last value of a field, one can overwrite any previously set transaction field. This is convenient for scripts that may wish to tweak values by appending lines to a transaction template.

Two special cases (for keys and assets) make the output easier for humans to process by providing compatibility with other tools.

The `._present` pseudoselector was selected to avoid conflicting with fields in the same data structure, since the XDR specification disallows fields that start with `_`. There is no similar ambiguity possible for `.len`, whose only siblings are bracketed array indices.

## Test Cases

The following binary transaction:

    AAAAACsWS5BDhC5BjpKQtznHFJ3CkU6+XtWopW+t+Q9KoH7QAAAAZAClKY0AAAABAAAAAQAAAABbicmAAAAAAF1q/QAAAAABAAAAFkVuam95IHRoaXMgdHJhbnNhY3Rpb24AAAAAAAEAAAAAAAAAAQAAAABAXzbt2M8i77+AcrmFtqTAFVHDTdOME3rI1A1ALNH3tAAAAAFVU0QAAAAAADJSVDIhkp9uz61Ra68rs3ScZIIgjT8ajX8Kkdc1be0LAAAAABfXk6AAAAAAAAAAAUqgftAAAABA3vtPH60cJ5MntVrxhP3N33P096jLQOflNKcdc6BRJLo2nbem0xtHyv0RhZIkaoV15sJJq5TsN2je22KSIhzlDA==

Can be rendered like this (note that comments are optional and may contain implementation-dependent information):

    tx.sourceAccount: GAVRMS4QIOCC4QMOSKILOOOHCSO4FEKOXZPNLKFFN6W7SD2KUB7NBPLN
    tx.fee: 100
    tx.seqNum: 46489056724385793
    tx.timeBounds._present: true
    tx.timeBounds.minTime: 1535756672 (Fri Aug 31 16:04:32 PDT 2018)
    tx.timeBounds.maxTime: 1567292672 (Sat Aug 31 16:04:32 PDT 2019)
    tx.memo.type: MEMO_TEXT
    tx.memo.text: "Enjoy this transaction"
    tx.operations.len: 1
    tx.operations[0].sourceAccount._present: false
    tx.operations[0].body.type: PAYMENT
    tx.operations[0].body.paymentOp.destination: GBAF6NXN3DHSF357QBZLTBNWUTABKUODJXJYYE32ZDKA2QBM2H33IK6O
    tx.operations[0].body.paymentOp.asset.type: ASSET_TYPE_CREDIT_ALPHANUM4
    tx.operations[0].body.paymentOp.asset.alphaNum4.assetCode: USD
    tx.operations[0].body.paymentOp.asset.alphaNum4.issuer: GAZFEVBSEGJJ63WPVVIWXLZLWN2JYZECECGT6GUNP4FJDVZVNXWQWMYI
    tx.operations[0].body.paymentOp.amount: 400004000 (40.0004e7)
    tx.ext.v: 0
    signatures.len: 1
    signatures[0].hint: 4aa07ed0 (GAVRMS4QIOCC4QMOSKILOOOHCSO4FEKOXZPNLKFFN6W7SD2KUB7NBPLN)
    signatures[0].signature: defb4f1fad1c279327b55af184fdcddf73f4f7a8cb40e7e534a71d73a05124ba369db7a6d31b47cafd118592246a8575e6c249ab94ec3768dedb6292221ce50c

## Implementation

Txrep is implemented by the Stellar transaction compiler, [stc](https://github.com/xdrpp/stc).

**Comment by stanford-scs** [comment]
@tomerweller What is the motivation for third-party tool support?  The proposed format is effectively already supported by any programming language's standard library, as every language lets you read input line-by-line and break a line at a colon character.  Something more complicated that requires a third-party library seems like net negative, particularly for processing of security-critical transactions.

How are you counting verbosity?  Your example requires more lines than the Key:Value format.  You are abbreviating the keys, but that comes at the cost of making the lines not self-describing and making the values order dependent.  You can't just append a line to a file to set a field's value, you have to find the right place in the file to add the line.

**Comment by tomerweller** [comment]
To encourage adoption, I think it's preferable to have a familiar syntax. With the above YAML, developers can choose to write their own parser, which should be straight forward, or use an existing library, which is something developer do these days.

True, the YAML has more lines than your syntax. But, the lines are shorter and with the nature of the size of operations, you are expected to have a good visibility into the hierarchy anyway.

I'm not religious about it, this spec is great either way. Just experiencing some syntax fatigue...

**Comment by nikhilsaraf** [comment]
This is a great SEP, I think we definitely need to standardize on a human-readable format for XDRs.

Here are some thoughts and suggestions:
- The spec should allow lines that only contain comments. A line that begins with `#` or `:` can be interpreted as a line that only has a comment.
- Fields — we should be explicit about using 0-indexed arrays for *all* arrays (not just variable length arrays).
- The fields should follow the XDR spec for variable names (i.e. not starting with uppercase letters)
- Consider using `.$present` and `.$len` instead of `.present` and `.len` since `present` and `len` can be valid fields on an XDR. This also has the benefit of allowing the format to be explicit about what is a field vs. what is a pseudo-selector.
- Strkey - the spec should provide links to libraries in popular languages. Here are a few:
    - [golang](https://github.com/stellar/go/blob/master/strkey/main.go)
    - [C++](https://github.com/stellar/stellar-core/blob/dd7f38dba232e20ae75a95641857c0b7ed0252e1/src/crypto/StrKey.cpp)
    - [JavaScript](https://github.com/stellar/js-stellar-base/blob/master/src/strkey.js).
- If possible we should include an ordering of values in the txrep file. This will bring about consistency in diffs between such files. We can use the ordering of the fields in the XDR definition.

**Comment by stanford-scs** [comment]
@nikhilsaraf Please have another look.  I believe I addressed all of your comments in the draft and stc implementation.

**Comment by nikhilsaraf** [comment]
@stanford-scs looks good, thanks for addressing all those comments. Excited to have this become a standard!

Found some possible typos, listing below:
- _Normalized txrep_ section — Fields in a strkey file -> Fields in a _txrep_ file
- _Rationale_ section - Pseudoselectors use '#' -> Pseudoselectors use `#`

**Comment by stanford-scs** [comment]
@nikhilsaraf Thanks.  Fixed typos.

**Comment by stanford-scs** [comment]
I'm starting to rethink the `#len` and `#present` pseudo-selector syntax.  It bugs me when looking at transactions.  We don't even need it for vectors, because `.len` won't conflict with `[0]` which is the sibling field.

The other thing is that it might be nice to extend this to arbitrary XDR, when debugging the ledger.  If ever we have a pointer to a pointer (which Stellar transactions do not currently have), the syntax will still be ambiguous.  Does `foo#present: true` mean `foo != NULL` or `*foo != NULL`?  I mean I don't know how many hoops it's worth jumping through to support pointers to pointers.  We can live without it, but if there happens to be a good solution that also accommodates that scenario, it would be better.

**Comment by orbitlens** [comment]
@tomerweller is right. Consider YAML syntax. It's one of the most popular and mature standards of human-readable structured data representation. There are plenty of well-tested and ready-to-use implementations of YAML parsers for all programming languages. Moreover, most developers have working experience with YAML as it becomes increasingly popular as a markup language for configuration files. I think that at least 20% of all projects on GitHub contain config files in YAML format.

I'd even suggested JSON as a most popular format, natively supported by JavaScript. It's important because JSON can be used directly in any browser without any additional coding or third-party libraries. Lack of comments is the only drawback I can see here. But do we really need comments that much?

Pseudo-selectors `#len` and `#present` are useless in most cases. Well, `#len` may be handy to display operations array length when it is needed (in case if a tx contains more than two operations, it may be hard to count them at a glance), but it can be specified as a comment. `#len` can be useful in the stream-based parsing for the precursive array memory allocation, but it will give slight performance boost only for large arrays (>1000), which is not the case.

`#present` is excessive; it does not improve readability, nor it is required for parsing, as the presence of an element can be determined implicitly following simple rules. For example, in the case of `time_bounds`, the structure may be considered as present only if either `min_time` or `max_time` is larger than `0`. Are there any other cases where the `#present` selector is essential?

**Comment by bartekn** [comment]
Just to add what previous commenters said: I also think that using YAML is a better option from a security standpoint. Writing a new parser is hard, it's very easy to make security related bugs. I remember that fuzzing revealed a lot of such bugs in our XDR libs (all of them already fixed).

It's just better to use already existing, well-tested YAML parsers.

**Comment by MisterTicot** [comment]
@orbitlens

I use a JSON format that can describe any Stellar transaction in cosmic-lib. It supports federated address and network field. You can play with it here: https://misterticot.github.io/js-cosmic-lib/demo.html?inflation

Tools are already available to parse it, convert it to cosmic links and XDR and display its HTML description. Maybe I should make into a SEP?

**Comment by orbitlens** [comment]
@MisterTicot I also use JSON representation in my parser internally. However, I must admit that YAML format is more laconic and human-readable, which is the primary goal of this SEP. Federated addresses and network selection is beyond the Transaction XDR data contract.

**Comment by MisterTicot** [comment]
@orbitlens Well you were the one suggesting a JSON solution... 🙄

While federated address and network are not part of XDR, it's definitely part of representing Stellar transactions.

**Comment by stanford-scs** [comment]
So YAML is pretty much a hard no.  As has been pointed out, writing parsers is hard.  Third-party parsers could contain bugs or open up implementations to attacks via package-management insecurities.  Hence, the format needs to be machine readable without a parser, or with just the rudimentary parsing logic that's already built into most programming languages, like sscanf, operator>>, fmt.Sscanf.  The current implementation needs only 15 lines of to parse a txrep format into field-value pairs, and has no dependencies other than golang's builtin `strings` package.

Also, an order-independent format is better, because it prevents excerpts from being misinterpreted out of context.  YAML is highly context-dependent.

The goal is to specify every possible transaction unambiguously, so the format needs to a human readable representation of XDR, not an approximate mapping between XDR and some other existing format.  Even if some transactions are not generally useful (like a TimeBound with all zeros), we need the ability to specify such a transaction, both for testing purposes and for preserving the validity of signatures (which will become invalid if we convert a meaningless TimeBound to a  NULL TimeBound*).  Any proposal in which XDR -> Txrep -> XDR is not guaranteed to return the exact same bytes for all possible `TransactionEnvelope` structures, even highly contrived ones, does not meet the txrep's goals.

**Comment by MisterTicot** [comment]
@stanford-scs I think JSON cosmic link format is close to that goal. Getting to a 100% xdr->json->xdr conversion is probably doable.

I would like to know your views about signatures? Should duplicate and invalide signatures be kept as well?

**Comment by stanford-scs** [comment]
JSON is hard to parse, generally requires some kind of third-party library that increases the attack surface, and, because it wasn't designed to provide a one-to-one mapping to XDR, will almost certainly lead to a bunch of underspecified cases that different implementations interpret differently.  Given that a txrep implementation already must parse binary XDR, you really don't want a whole second parser with potentially different logic.  Txrep should be parsed by auto-generated XDR code.  Most implementations abstract over some `XDR` interface to select either serializing or deserializing binary XDR; you can just leverage that to create two more `XDR` implementations that go to/from txrep (or in the from case, from a field-value map trivially parsed from a txrep file).

The point of formats such as JSON is to provide a native data representation and parser for external data.  When you already have a native representation and parser generated by an XDR compiler, there's really no need for JSON.  In particular, if XDR were an ASCII format, we wouldn't need txrep at all--we would just look at the XDR.  Since it isn't, txrep is essentially providing an alternate ASCII format that can be parsed by existing XDR implementations given a simple new back-end implementation.

That's not to say there isn't value in representing Stellar transactions as JSON, so as to view or manipulate them in code that doesn't contain an XDR-compiled `TransactionEnvelope` structure.  That just isn't the goal of txrep.  Txrep is for humans to be able to understand and specify any bytes that could be parsed as an XDR `TransactionEnvelope`.  That includes ones with bad or duplicate signatures. For example, if some transaction causes your software to crash, you should be able to decompile it to txrep and immediately see what is going on.  You should also be able to specify invalid transactions for regression tests or higher-layer fuzz testing.

**Comment by MisterTicot** [comment]
@stanford-scs Thank you for that detailed answer, I now understand better what you want to achieve.

**Comment by stanford-scs** [comment]
Now I'm thinking of pseudoselectors `.len` for length and either `.present?` or just `?` for pointers.  The question mark is makes intuitive sense, but also will prevent any name clashes.  Any reactions?

I don't think it's worth jumping through too many hoops for pointers-to-pointers that don't currently exist in the protocol, so long as there is some potential way to deal with it in the future.  Maybe the answer is as simple as adding a third pseudoselector (maybe `*`) to represent a pointer within a pointer.  So if field `foo` really is a pointer to pointer, you could have `foo.present?: true` and `foo*.present?: false` to represent a non-NULL pointer to a NULL pointer.  That's ugly because most languages use a prefix operator to dereference pointers, but again this isn't something super likely to come up.

**Comment by stanford-scs** [comment]
Okay, I've revised the proposal yet again to use a more readable but still unambiguous pseudoselector syntax of `.present?` and `.len`, as well as defined syntax that will continue to be unambiguous in the unlikely event that some future operation requires pointers-to-pointers.

**Comment by stanford-scs** [comment]
XDR disallows identifiers that start with underscore <https://tools.ietf.org/html/rfc4506#section-6.2>.  Should I just use `_present` instead of  `present?`.

**Comment by stanford-scs** [comment]
@orbitlens @tomerweller So I did think more about YAML tonight, since you guys feel so strongly about it.  Specifically, I looked into what it would take to implement something like [stc](https://github.com/xdrpp/stc) using YAML.

The first problem that became immediately apparent is that YAML is not XDR.  As the YAML spec states

> YAML is both a text format and a method for presenting any native data structure in this format. Therefore, this specification defines two concepts: a class of data objects called YAML representations, and a syntax for presenting YAML representations as a series of characters, called a YAML stream.

XDR, too, provides both data object representations and serialization, so the question is whether the two data object representations can be made compatible, and I'm pretty sure the answer is not easily.

Since stc is in go, I looked at several go YAML parsers based on [this blog post](http://sweetohm.net/article/go-yaml-parsers.en.html).  None of them is capable of parsing YAML into the go data structures produced by `goxdr` (or probably any XDR compiler, for that matter).  A number of the YAML libraries just output hierarchical tree-like structures.  Go-yaml will parse into a go data structure, but would have no way of parsing XDR unions (or enforcing maximum string lengths, or parsing enums, etc.).

Essentially, to avoid manually translating between two different go data structures both representing a `TransactionEnvelope`, you would need to parse the YAML and turn it into something conceptually equivalent to txrep.  So this would be a pretty big net negative on the software complexity side--add a big dependency while also needing to implement extra code to flatten out a YAML tree for easy lookup by an XDR parser. If you believe this is wrong, I'd invite you to try it--grab [xdr_generated.go](https://raw.githubusercontent.com/xdrpp/stc/go1/xdr_generated.go) and see if you can get a YAML library to parse into a `TransactionEnvelope` or even a much simpler union type such as `Memo`.

The remaining argument is that YAML could be  easier for humans to read, and that this readability is worth the extra implementation complexity.  There are probably some use cases where it is appropriate to translate Stellar transactions into YAML, just as there are reasons to represent a transaction with JSON. 
 The existence of txrep shouldn't preclude such representations.  But the [YAML spec](http://yaml.org/spec/1.2/spec.html) is considerably longer and more complex than txrep, and leaves a lot of ambiguities.  For example, whether or not a YAML file may contain 64-bit integers is [implementation dependent](http://yaml.org/spec/1.2/spec.html#id2803828)--so does that mean we need to use strings every place a Stellar transaction contains a 64-bit number?  How should a program interpret fields that have the wrong type, or unions that have the wrong field, or enums with non-existent constants?  Again, it seems you are going to have to reduce things conceptually to something like the existing txrep to explain them.  Basically you'll say that inside tx, inside timeBound, the field maxBound must have the same type as the C/C++/go field `tx.timeBound.maxBound`.  So why not just use the C notation?

So while my main objection to YAML is the increase in implementation complexity and attack surface, I think the human factor is at best a wash, and YAML would certainly not be categorically better for all applications.  If someone cares enough to implement it and then still thinks YAML a good idea, I guess we can talk more, but it looks highly non-trivial to me.  In fact, the easiest implementation would be YAML -> txrep -> XDR.  So I think the best think to do is standardize txrep, and if someone cares enough to do it later on, also have a JSON or YAML convention (which should be called JSON or YAML, respectively, instead of txrep).

**Comment by MisterTicot** [comment]
@stanford-scs

Maybe the problem lies in the fact your trying to solve two different problems at once:

* Having a human syntax for XDR
* Having a standard way to represent TX

Both are of course closely related. However, while I could use your syntax into a program like a smart contract generator, I would certainly avoid to use it in documentation and such. This is because the syntax is kind of bloated and repeat itself very much, which is not so great when you just want to explain something.

In the first case exact translation is more important than syntax clarity. In the second case it's the opposite

Maybe it could be a good idea to consider an alternative easier syntax for cases were the transaction representation is not going to be directly parsed, but is here for explanatory purpose.

The two format could be related and eventually presented in the same SEP.

I presented a [SEP draft](https://github.com/stellar/stellar-protocol/blob/19525dbd23256bb5803c4864266f353a581ee057/drafts/sep-0011.md) were I could make use of such a format. I had to specify the network, by the way ;) @orbitlens

**Comment by stanford-scs** [comment]
@MisterTicot I think something like what you propose is somewhat complementary to txrep.  In particular, the draft you link to could be strengthened through concrete test cases, and it would improve readability to be able to specify the output in txrep in addition to or instead of base64-encoded binary XDR.

Maybe the way to think of txrep is as a kind of assembly language for XDR.  It has to be human readable and should be _a_ standard, but should not preclude higher-level languages becoming standard as well.  If anything, txrep should facilitate higher-layer languages because now you can specify and debug your compilers more easily.

So these points could be addressed by tweaking motivation.

**Comment by MisterTicot** [comment]
@stanford-scs

Ok I guess I misunderstood the meaning of "describing transactions" the first paragraph of motivation.

**Comment by nikhilsaraf** [comment]
@stanford-scs either `_present` or `present?` would work for me.
Also like the idea of keeping the current format as the lowest-layer of a serialized format and also allowing additional formats (JSON/YAML) on top of that too.
Ship it! :ship:

**Comment by stanford-scs** [comment]
I don't know how to turn an issue into a pull request, so I created a new pr #169.  It seems you need to download some tool or something and can't do it on the web site or just with git.

Anyway, please continue discussion at #169 .
